swap_meet/vendor.py

```diff
@@ -1,2 +1,157 @@
+    #  
+
+    # def attr_decorator(attr, func):
+    #     def a_decorator_passing_arguments(method_to_decorate):
+    #             def a_wrapper_accepting_arg(self, category):
+    #                 matching_category=self.get_by_category(category)
+    #                 return  func(matching_category, key =attrgetter(attr), default=None)
+    #             return a_wrapper_accepting_arg
+    #     return a_decorator_passing_arguments          
+
+    # @attr_decorator(attr='condition', func = max)
+    # def get_best_by_category(self, category):
+    #     pass
+    
+    # @attr_decorator(attr='age', func = min)
+    # def get_best_by_age(self, category):
+    #     pass
+    
+
+    # def attr_decorator(func):
+    #     def a_decorator_passing_arguments(method_to_decorate):           
+    #         def a_wrapper_accepting_arg(self,other,my_priority, their_priority):
+    #                 my_best_item=self.func(self, their_priority)
+    #                 their_best_item=other.func(other, my_priority)
+    #                 if my_best_item is None or their_best_item is None:
+    #                     return False
+    #                 self.swap_items(other, my_best_item, their_best_item)
+    #                 return True         
+    #         return a_wrapper_accepting_arg
+    #     return a_decorator_passing_arguments        
+
+    # @attr_decorator(func =get_best_by_category)
+    # def swap_best_by_category(self,other,my_priority, their_priority):
+    #     return True
+
+    # @attr_decorator(func =get_best_by_age)
+    # def swap_by_newest(self,other,my_priority, their_priority):
+    #     return True    
+
+    #* the code above should be refactored to this:
+
+
+    def get_best_by(self, category, attr, func):
+        matching_category=self.get_by_category(category)
+        res = func(matching_category, key =attrgetter(attr), default=None)
+        return res         
+
+    def get_best_by_category(self, category):
+        return self.get_best_by(category, 'condition', max)
+    
+    def get_best_by_age(self, category):
+        return self.get_best_by(category, 'age', min)
+
+    func={'category':get_best_by_category, 'age':get_best_by_age}
+    
+    def swap_best_by(self,other,my_priority, their_priority, f):
+            my_best_item=self.func[f](self,their_priority)
+            their_best_item=other.func[f](other, my_priority)
+            if my_best_item is None or their_best_item is None:
+                return False
+            self.swap_items(other, my_best_item, their_best_item)
+            return True         
+
+    def swap_best_by_category(self, other, my_priority, their_priority):
+        return  self.swap_best_by(other, my_priority, their_priority, "category")
+
+    def swap_by_newest(self,other, my_priority, their_priority):
```

Comment:
Great work!  With a few small modifications this could also be used to create a `swap_by_newest` that does not include the category (so the two newest would swap regardless of category).  One implementation could be to remove priorities from swap_by_newest, pass `None` to swap_best_by, and have get_by_category returned the full inventory when the category is None.